File: 2023/day01/src/main.py
import os
current_dir = os.path.dirname(os.path.abspath(__file__))
with open(os.path.join(current_dir, 'input.txt')) as f:
    with open(os.path.join(current_dir, 'results_rust.txt')) as r:
        lines = f.readlines()
        results = r.readlines()
        numbers = {
            "one": "1",
            "two": "2",
            "three": "3",
            "four": "4",
            "five": "5",
            "six": "6",
            "seven": "7",
            "eight": "8",
            "nine": "9",
        }
        calibration = 0
        i = 0
        for line in lines:
            first_index = len(line)-1
            first = line[first_index]
            last_index = 0
            last = line[last_index]
            for number_key, number in zip(numbers.keys(), numbers.values()):
                if number_key in line:
                    if line.index(number_key) < first_index:
                        first_index = line.index(number_key)
                        first = numbers[number_key]
                    if line.rindex(number_key) > last_index:
                        last_index = line.rindex(number_key)
                        last = numbers[number_key]
                if number in line:
                    if line.index(number) < first_index:
                        first_index = line.index(number)
                        first = number
                    if line.rindex(number) > last_index:
                        last_index = line.rindex(number)
                        last = number
            # print(f"{first}{last}")
            if f"{first}{last}" != results[i].strip():
                print(f"{i}: {first}{last} != {results[i].strip()}")
            calibration += int(first + last)
            i += 1
        print(calibration)

File: 2023/day01/src/main.rs
use aho_corasick::{AhoCorasick, MatchKind};
use std::collections::HashMap;

fn main() {
    let input = include_str!("input.txt");
    part_one(input);
    part_two(input);
}

fn part_one(input: &str) {
    let mut calibration: i32 = 0;
    for line in input.lines() {
        let calibration_values: Vec<char> = line.chars().filter(|c| c.is_digit(10)).collect();

        if calibration_values.len() > 0 {
            let number = format!("{}{}", calibration_values[0], calibration_values[calibration_values.len() - 1]);
            calibration += number.parse::<i32>().unwrap();
        }
    }
    print!("Calibration part one: {}\n", calibration)
}

fn part_two(input: &str) {
    let mut calibration: i32 = 0;
    let numbers: HashMap<&str, &str> = HashMap::from([("one", "1"), ("two", "2"), ("three", "3"), ("four", "4"), ("five", "5"), ("six", "6"), ("seven", "7"), ("eight", "8"), ("nine", "9")]);
    let all = numbers.keys().chain(numbers.values()).collect::<Vec<_>>();
    // let ac = AhoCorasick::new(all.clone()).unwrap();
    let ac = AhoCorasick::builder()
        .match_kind(MatchKind::Standard)
        .build(all.clone()).unwrap();

    for line in input.lines() {
        let mut matches = vec![];
        for mat in ac.find_overlapping_iter(line) {
            matches.push((mat.pattern(), mat.start(), mat.end()));
        }
        let mut calibration_value = "".to_owned();
        if let Some(min_match) = matches.iter().min_by_key(|&(_, start, _)| start) {
            let num = all[min_match.0.as_usize()];
            if numbers.contains_key(num) {
                calibration_value.push_str(numbers[num])
            } else {
                calibration_value.push_str(num)
            }
        }
        if let Some(max_match) = matches.iter().max_by_key(|&(_, _, end)| end) {
            let num = all[max_match.0.as_usize()];
            if numbers.contains_key(num) {
                calibration_value.push_str(numbers[num])
            } else {
                calibration_value.push_str(num)
            }
        }
        calibration += calibration_value.parse::<i32>().unwrap();
        
    }
    print!("Calibration part two: {}\n", calibration)
}
